lib/Codegen/Codegen.cpp

```diff
@@ -99,6 +99,22 @@ llvm::Value *Codegen::getValue(Inst *I) {
   if (!V0)
     return nullptr;
 
+  // PHI nodes must be the first instructions in a basic block. If we're
+  // replacing a PHI node with another instruction, make sure it comes after the
+  // other PHI nodes.
+  if (ReplacedInst &&
+      !ReplacedInst->isTerminator() &&
+      isa<PHINode>(ReplacedInst->getNextNode()) &&
+      I->K != Inst::Kind::Phi) {
+    Instruction *InsertPoint = ReplacedInst;
+    while (!InsertPoint->isTerminator() &&
+           isa<PHINode>(InsertPoint->getNextNode()))
+      InsertPoint = InsertPoint->getNextNode();
+    if (InsertPoint->isTerminator())
```

Comment:
Hi @zoecarver , 

Thanks for the patch! One concern:  as stated in [1], after calling IRBuilder.SetInsertPoint(Instruction *I), the new instructions will be inserted *before* instruction I. In this case, InsertPoint should be the instruction just next to the last phi in that basic block, right? However, this code would make InsertPoint the last PHI node and Instruction could still possibly be inserted between PHI nodes.

https://llvm.org/doxygen/classllvm_1_1IRBuilderBase.html#ae72654261eee259da46249dc919b029c

Reply:
You're right. Fixed.

Reply:
Since IR builder is inserting before the ReplacedInst, it should work even when ReplacedInst is a terminator, right? Therefore, checking the terminator is not necessary here. 

And Souper won't create Phis, the check ```I->K != Inst::Kind::Phi``` is not necessary here.

I think 

```
if (ReplacedInst) {
  Instruction *InsertPoint = ReplacedInst;
  while (isa<PHINode>(InsertPoint) {
    InsertPoint = InsertPoint->getNextNode();
  }
  Builder.SetInsertPoint(InsertPoint);
}
```

should just do the work.

Reply:
Good point. I'll update it with your suggestion. I think it would be good to keep the `isa<PHINode>(ReplacedInst->getNextNode())` check so that we don't unnecessarily update the `InsertPoint`, though.